### DISCUSSION TITLE
How to auto completion the brackets and arguments for function?

### DISCUSSION THREAD

**Discussion by csmajorone** [hrsh7th/nvim-cmp#1022] [Q&A]
I use an online dotfile and find that the function brackets and arguments can be auto-completed (after `<Tab>` or `<Enter>`), like this:
![function_with](https://user-images.githubusercontent.com/104993836/171790584-611e0076-9b96-401d-ac81-6a043e29464b.gif)

I basically copied its `config/cmp.lua` file as my own, but they are not autocompleted (After `<Enter>` to confirm the selection),
![function_without](https://user-images.githubusercontent.com/104993836/171790780-f9ff8fd6-a25d-4016-bcc7-ea39bc90e990.gif)

So I was wondering, what could be the problem? what is the switch to control that?

**Comment by Shougo** [comment]
snippet plugin installation is needed for it.

https://github.com/hrsh7th/nvim-cmp#recommended-configuration

**Reply by csmajorone** [reply]
The source is LSP, not from snippet.

**Reply by Shougo** [reply]
LSP uses snippet feature.  It is used in your desired argument completion.

**Reply by csmajorone** [reply]
I don't know that, but can you be more specific? what argument exactly?

**Reply by Shougo** [reply]
Please see this.

https://github.com/microsoft/language-server-protocol/blob/main/snippetSyntax.md

**Reply by csmajorone** [reply]
I don't think it can solve my problem, maybe you just guess something, and a protocol is different from a specific implementation. As I said before, I use the same `cmp.lua`, and the snippet source is the same.

**Reply by Shougo** [reply]
I need to guess, because you don't upload the minimal init.vim.
Please upload the minimal init.vim.  I will check it later.

**Reply by Shougo** [reply]
Please see this.

https://github.com/hrsh7th/nvim-cmp/issues/852

And it works for me.

```vim
lua << EOF
-- Add additional capabilities supported by nvim-cmp
local capabilities = vim.lsp.protocol.make_client_capabilities()
capabilities = require('cmp_nvim_lsp').update_capabilities(capabilities)

require'lspconfig'.gopls.setup{
  capabilities = capabilities,
  init_options = {
    usePlaceholders = true,
  },
}
EOF
```

**Reply by csmajorone** [reply]
My config is a bit entangled, so it takes time to chop down to a minimal, but I did set `usePlaceholders = true` in my gopls. 

```
local opts = {
  init_options = {
    usePlaceholders = true,
  },
  settings = {
    gopls = {
      allExperiments = true,
      usePlaceholders = true,
      analyses = {
        nilness = true,
        shadow = true,
        unusedparams = true,
        shadow = true,
     },
     staticcheck = true,
    },
  },
  flags = {
    debounce_text_changes = 150,
  },
  cmd = { "gopls" },
  filetypes = {"go", "gomod", "gotmpl"},
  single_file_support = true,
}
```

**Reply by Shougo** [reply]
So it seems configuration problem.
You need to create minimal config.

**Reply by csmajorone** [reply]
can you check [this dotfile](https://github.com/akinsho/dotfiles)? I use it, and it has this effect:
![function_with](https://user-images.githubusercontent.com/104993836/171790584-611e0076-9b96-401d-ac81-6a043e29464b.gif)

I copied its `cmp.lua`,  but unfortunately, it hehaves differently when working with my configs, I don't know why.

**Reply by hrsh7th** [reply]
Essentially, you need to look for a minimal reproduction configuration to raise the issue.
At the very least, we don't know anything without pasting your settings.

**Reply by Shougo** [reply]
@csmajorone I don't want to test your whole dotfiles.  It is too long.
Please test the miinmal vimrc instead.

```vim
set rtp+=~/src/nvim-cmp
set rtp+=~/src/cmp-buffer
set rtp+=~/src/cmp-nvim-lsp
set rtp+=~/src/vim-vsnip
set rtp+=~/src/cmp-vsnip
set rtp+=~/src/nvim-lspconfig
set rtp+=~/src/cmp-cmdline

lua <<EOF
  -- Setup nvim-cmp.
  local cmp = require'cmp'

  cmp.setup({
    snippet = {
      expand = function(args)
        print(args)
        vim.fn["vsnip#anonymous"](args.body)
      end,
    },
    mapping = {
      ['<CR>'] = cmp.mapping.confirm {
        behavior = cmp.ConfirmBehavior.Replace,
        select = true,
      },
      ['<tab>'] = cmp.mapping(cmp.mapping.confirm({ select = true }), { 'i', 's', 'c' }),
      ['<C-n>'] = cmp.mapping({
          i = function(fallback)
              if cmp.visible() then
                  cmp.select_next_item({ behavior = cmp.SelectBehavior.Select })
              else
                  fallback()
              end
          end
      }),
      ['<C-p>'] = cmp.mapping({
          i = function(fallback)
              if cmp.visible() then
                  cmp.select_prev_item({ behavior = cmp.SelectBehavior.Select })
              else
                  fallback()
              end
          end
      }),
    },
    sources = {
      { name = 'nvim_lsp' },
      { name = 'buffer' },
    },
    completion = {
    completeopt = 'menu,menuone,noselect'
    }
  })
  cmp.setup.cmdline(':', {
    mapping = cmp.mapping.preset.cmdline(),
    sources = {
      { name = 'cmdline' }
    }
  })
EOF

lua << EOF
-- Add additional capabilities supported by nvim-cmp
local capabilities = vim.lsp.protocol.make_client_capabilities()
capabilities = require('cmp_nvim_lsp').update_capabilities(capabilities)

--require'lspconfig'.sumneko_lua.setup{}
--require'lspconfig'.cssls.setup{}
--require'lspconfig'.pylsp.setup{}
require'lspconfig'.gopls.setup{
  capabilities = capabilities,
  init_options = {
    usePlaceholders = true,
  },
}
EOF

runtime! after/plugin/*.lua
```

**Reply by Shougo** [reply]
OK.  I get the reason why your config does not work.  You need to initialize capabilities like this.

```lua
local capabilities = vim.lsp.protocol.make_client_capabilities()
capabilities = require('cmp_nvim_lsp').update_capabilities(capabilities)

require'lspconfig'.gopls.setup{
  capabilities = capabilities,
  init_options = {
    usePlaceholders = true,
  },
  settings = {
    gopls = {
      allExperiments = true,
      usePlaceholders = true,
      analyses = {
        nilness = true,
        shadow = true,
        unusedparams = true,
        shadow = true,
     },
     staticcheck = true,
    },
  },
  flags = {
    debounce_text_changes = 150,
  },
  cmd = { "gopls" },
  filetypes = {"go", "gomod", "gotmpl"},
  single_file_support = true,
}
```

It works for me.

**Reply by Shougo** [reply]
Hm...  LSP configuration is very complex.

**Reply by csmajorone** [reply]
Do you know what `capabilities` exactly for this effect? I spent tons of hours trying to debug my configs, still not working. 
Also I think the order of lazy-loaded plugins caused some of my problems that `require'cmp_nvim_lua'.update_capabilities` failed.

**Reply by csmajorone** [reply]
After I removed all lazy-loaded stuff (which caused `require'cmp_nvim_lua'.update_capabilities` to fail), I can succeed with the effect. I agree, lspconfig is so complicated, I kinda appreciate the effort from **helix**, which makes lsp working out of the box.

**Reply by csmajorone** [reply]
Alright, I checked `cmp_nvim_lua`'s source code, and it's actually `capabilities.textDocument.completion.completionItem.snippetSupport` property makes play.
